db/migrate/20170703094235_add_hash_expression_to_miq_alert.rb

```diff
@@ -0,0 +1,35 @@
+class AddHashExpressionToMiqAlert < ActiveRecord::Migration[5.0]
+  class MiqAlert < ActiveRecord::Base
+    serialize :expression
+    serialize :miq_expression
+    serialize :hash_expression
+  end
+
+  def up
+    add_column :miq_alerts, :hash_expression, :text
+
+    say_with_time('Add hash expression to MiqAlert for existing alerts') do
+      MiqAlert.all.each do |alert|
+        next unless alert.expression.kind_of?(Hash)
+        alert.hash_expression = alert.expression
+        alert.expression = nil
+        alert.save
+      end
+    end
```

Comment:
Reading this PR carefully, it seems you only care about when the type of the expression is a hash, so to avoid deserializing MiqExpression, you can use string comparisons on the up, and then just transfer the column normally.

- Remove the serialize lines from the MiqAlert stub
- Use string comparisons looking for MiqExpression types in the columns
  ```ruby
  say_with_time('Add hash expression to MiqAlert for existing alerts') do
    MiqAlert.all.each do |alert|
      next if alert.expression.start_with?("--- !ruby/object:MiqExpression")
      alert.hash_expression = alert.expression
      alert.expression = nil
      alert.save
    end
  end
  ```
- The down can actually stay the same.

---

This has the side effect of making the PR faster because you don't go through `deserialization -> type check -> serialize`, but instead just do `string check -> copy`

---

Taking this a step further, since you know the anti-signature, you can do the filtering at the query level with something like

```ruby
MiqAlert.where("expression NOT LIKE ?", "--- !ruby/object:MiqExpression%").each do |alert|
  alert.hash_expression = alert.expression
  alert.expression = nil
  alert.save
end
```

Reply:
@Fryguy I made the changes, thanks!